File: models/Parent.js
const mongoose = require("mongoose")
const bcrypt = require("bcryptjs")

const parentSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: true,
      trim: true,
    },
    email: {
      type: String,
      required: true,
      unique: true,
      trim: true,
      lowercase: true,
      match: [/^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/, "Please enter a valid email"],
    },
    password: {
      type: String,
      required: true,
      minlength: 6,
    },
    children: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "User",
        default: [], // Add default empty array
      },
    ],
    isActive: {
      type: Boolean,
      default: true,
    },
    isTemporaryPassword: {
      type: Boolean,
      default: false,
    },
    lastLoginAt: {
      type: Date,
      default: null,
    },
    passwordResetToken: {
      type: String,
      default: null,
    },
    passwordResetExpires: {
      type: Date,
      default: null,
    },
    // Blocking system
    isBlocked: {
      type: Boolean,
      default: false,
    },
    blockInfo: {
      reason: {
        type: String,
        trim: true,
        default: "",
      },
      notes: {
        type: String,
        trim: true,
        default: "",
      },
      blockedAt: {
        type: Date,
        default: null,
      },
      blockedBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: "Admin",
        default: null,
      },
      expiresAt: {
        type: Date,
        default: null,
      },
      autoUnblockProcessed: {
        type: Boolean,
        default: false,
      },
    },
    paymentMethod: {
      type: String,
      enum: ["bank_transfer", "sepa_direct_debit"],
      default: null,
    },
    paymentDetails: {
      bankDetails: {
        accountHolderName: { type: String, trim: true },
        bankName: { type: String, trim: true },
        accountNumber: { type: String, trim: true },
        routingNumber: { type: String, trim: true },
        iban: { type: String, trim: true },
        swiftCode: { type: String, trim: true },
      },
      sepaDetails: {
        accountHolderName: { type: String, trim: true },
        iban: { type: String, trim: true },
        bic: { type: String, trim: true },
        mandateReference: { type: String, trim: true },
        mandateDate: { type: Date },
      },
    },
    paymentPreferences: {
      autoDebit: { type: Boolean, default: false },
      monthlyLimit: { type: Number, min: 0, default: 0 },
      notifications: { type: Boolean, default: true },
    },
  },
  {
    timestamps: true,
  },
)

// Index for better query performance
parentSchema.index({ email: 1 })
parentSchema.index({ createdAt: -1 })

// Hash password before saving
parentSchema.pre("save", async function (next) {
  // Only hash the password if it has been modified (or is new)
  if (!this.isModified("password")) return next()

  try {
    // Hash password with cost of 12
    const hashedPassword = await bcrypt.hash(this.password, 12)
    this.password = hashedPassword
    next()
  } catch (error) {
    next(error)
  }
})

// Method to check password
parentSchema.methods.checkPassword = async function (candidatePassword) {
  return await bcrypt.compare(candidatePassword, this.password)
}

// Method to add child to parent
parentSchema.methods.addChild = function (childId) {
  // Ensure children array exists
  if (!this.children) {
    this.children = []
  }

  // Check if child is already linked
  if (!this.children.includes(childId)) {
    this.children.push(childId)
    return this.save()
  }
  return Promise.resolve(this)
}

// Method to remove child from parent
parentSchema.methods.removeChild = function (childId) {
  // Ensure children array exists
  if (!this.children) {
    this.children = []
    return this.save()
  }

  this.children = this.children.filter((child) => !child.equals(childId))
  return this.save()
}

// Method to update last login
parentSchema.methods.updateLastLogin = function () {
  this.lastLoginAt = new Date()
  return this.save()
}

// Virtual for children count with null check
parentSchema.virtual("childrenCount").get(function () {
  return this.children ? this.children.length : 0
})

// Ensure virtual fields are serialized
parentSchema.set("toJSON", { virtuals: true })

module.exports = mongoose.model("Parent", parentSchema)

File: flaskrfid/models/parent.py
from datetime import datetime
from mongoengine import Document, StringField, BooleanField, DateTimeField, ListField, ReferenceField, EmbeddedDocument, EmbeddedDocumentField, FloatField
import bcrypt

class BlockInfo(EmbeddedDocument):
    reason = StringField(default='')
    notes = StringField(default='')
    blocked_at = DateTimeField()
    blocked_by = ReferenceField('Admin')
    expires_at = DateTimeField()
    auto_unblock_processed = BooleanField(default=False)

class BankDetails(EmbeddedDocument):
    account_holder_name = StringField()
    bank_name = StringField()
    account_number = StringField()
    routing_number = StringField()
    iban = StringField()
    swift_code = StringField()

class SepaDetails(EmbeddedDocument):
    account_holder_name = StringField()
    iban = StringField()
    bic = StringField()
    mandate_reference = StringField()
    mandate_date = DateTimeField()

class PaymentDetails(EmbeddedDocument):
    bank_details = EmbeddedDocumentField(BankDetails)
    sepa_details = EmbeddedDocumentField(SepaDetails)

class PaymentPreferences(EmbeddedDocument):
    auto_debit = BooleanField(default=False)
    monthly_limit = FloatField(min_value=0, default=0)
    notifications = BooleanField(default=True)

class Parent(Document):
    name = StringField(required=True)
    email = StringField(required=True, unique=True)
    password = StringField(required=True, min_length=6)
    children = ListField(ReferenceField('User'), default=list)
    is_active = BooleanField(default=True)
    is_temporary_password = BooleanField(default=False)
    last_login_at = DateTimeField()
    password_reset_token = StringField()
    password_reset_expires = DateTimeField()
    is_blocked = BooleanField(default=False)
    block_info = EmbeddedDocumentField(BlockInfo, default=BlockInfo)
    payment_method = StringField(choices=['bank_transfer', 'sepa_direct_debit'])
    payment_details = EmbeddedDocumentField(PaymentDetails)
    payment_preferences = EmbeddedDocumentField(PaymentPreferences, default=PaymentPreferences)
    created_at = DateTimeField(default=datetime.utcnow)
    updated_at = DateTimeField(default=datetime.utcnow)
    
    meta = {
        'collection': 'parents',
        'indexes': [
            'email',
            '-created_at'
        ]
    }
    
    def clean(self):
        """Clean and validate data before saving"""
        if self.email:
            self.email = self.email.lower().strip()
        self.updated_at = datetime.utcnow()
        
        # Hash password if it's been modified
        if self._changed_fields and 'password' in self._changed_fields:
            self.password = bcrypt.hashpw(self.password.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
    
    def check_password(self, candidate_password):
        """Check if provided password matches stored hash"""
        return bcrypt.checkpw(candidate_password.encode('utf-8'), self.password.encode('utf-8'))
    
    def add_child(self, child_id):
        """Add child to parent"""
        if child_id not in self.children:
            self.children.append(child_id)
            self.save()
        return self
    
    def remove_child(self, child_id):
        """Remove child from parent"""
        if child_id in self.children:
            self.children.remove(child_id)
            self.save()
        return self
    
    def update_last_login(self):
        """Update last login timestamp"""
        self.last_login_at = datetime.utcnow()
        self.save()
        return self
    
    @property
    def children_count(self):
        """Get number of children"""
        return len(self.children) if self.children else 0
    
    def to_dict(self):
        """Convert to dictionary for JSON serialization"""
        return {
            'id': str(self.id),
            'name': self.name,
            'email': self.email,
            'children_count': self.children_count,
            'is_active': self.is_active,
            'is_temporary_password': self.is_temporary_password,
            'last_login_at': self.last_login_at.isoformat() if self.last_login_at else None,
            'is_blocked': self.is_blocked,
            'payment_method': self.payment_method,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }